Add default executionControl reconciliation for common Maven plugins

cowwoc opened the pull request:
## Problem

Without executionControl configuration, the build cache does not track critical plugin properties that are often specified via command-line arguments (e.g., `-Dmaven.compiler.release`). This leads to incorrect cache reuse when these properties change between builds.

In multi-module JPMS projects, this manifests as compilation failures or bytecode version mismatches:

1. Build with `-Dmaven.compiler.release=17` → cache stores Java 17 bytecode (major version 61)
2. Build with `-Dmaven.compiler.release=21` → cache incorrectly reuses Java 17 bytecode  
3. Result: Bytecode remains major version 61 instead of expected 65

This is particularly problematic for `module-info.class` files which are sensitive to Java version.

## Solution

Implemented default reconciliation configs for common plugins when executionControl is not specified:

- **maven-compiler-plugin** (`compile` & `testCompile` goals):
  - Tracks `source`, `target`, and `release` properties
  - Ensures cache invalidation when Java version changes
  
- **maven-install-plugin** (`install` goal):
  - Tracked to ensure local repository is updated when needed

Users can still provide explicit `executionControl` configuration which will override these defaults.

## Testing

Verified with multi-module JPMS test project. See commit message for detailed test results.

**Before (broken):**
```bash
mvn clean verify -Dmaven.compiler.release=17  # Caches Java 17 bytecode
mvn clean verify -Dmaven.compiler.release=21  # Incorrectly reuses Java 17 bytecode
javap -v module-info.class | grep "major version"  # Shows 61 (wrong!)
```

**After (fixed):**
```bash
mvn clean verify -Dmaven.compiler.release=17  # Caches Java 17 bytecode  
mvn clean verify -Dmaven.compiler.release=21  # Detects change, recompiles
javap -v module-info.class | grep "major version"  # Shows 65 (correct!)
```

## Impact

- ✅ Users no longer need to manually configure executionControl for basic scenarios
- ✅ Prevents silent bytecode version mismatches in JPMS projects
- ✅ Backward compatible: explicit executionControl config still takes precedence

## Related Issues

This PR addresses multiple user-reported issues:

- **#373** - Support Java Modules and Multi-Module Projects out of the box (directly addresses the request for default compiler reconciliation)
- **#375** - Maven Build Cache Extension incorrectly handles --module-version compiler arguments (related compiler argument tracking issue)
- **#268** - Add project version as additional property for reconciliation (similar tracking requirement)
- **#272** - Allow to define 'reconciles' at plugin execution level (related reconciliation configuration)

cowwoc committed:
Add default executionControl reconciliation for common Maven plugins
## Problem
Without executionControl configuration, the build cache does not track critical plugin properties that are often specified via command-line arguments (e.g., -Dmaven.compiler.release). This leads to incorrect cache reuse when these properties change between builds.
In multi-module JPMS projects, this manifests as compilation failures or bytecode version mismatches:
1. Build with `-Dmaven.compiler.release=17` → cache stores Java 17 bytecode (major version 61)
2. Build with `-Dmaven.compiler.release=21` → cache incorrectly reuses Java 17 bytecode
3. Result: Bytecode remains major version 61 instead of expected 65
This is particularly problematic for module-info.class files which are sensitive to Java version.
## Solution
Implemented default reconciliation configs for common plugins when executionControl is not specified:
- **maven-compiler-plugin** (compile & testCompile goals):
  - Tracks `source`, `target`, and `release` properties
  - Ensures cache invalidation when Java version changes
- **maven-install-plugin** (install goal):
  - Tracked to ensure local repository is updated when needed
## Testing
Verified with multi-module JPMS test project:
**Before (broken):**
```
mvn clean verify -Dmaven.compiler.release=17  # Caches Java 17 bytecode
mvn clean verify -Dmaven.compiler.release=21  # Incorrectly reuses Java 17 bytecode
javap -v module-info.class | grep "major version"  # Shows 61 (wrong!)
```
**After (fixed):**
```
mvn clean verify -Dmaven.compiler.release=17  # Caches Java 17 bytecode
mvn clean verify -Dmaven.compiler.release=21  # Detects change, recompiles
javap -v module-info.class | grep "major version"  # Shows 65 (correct!)
```
## Impact
- Users no longer need to manually configure executionControl for basic scenarios
- Prevents silent bytecode version mismatches in JPMS projects
- Backward compatible: explicit executionControl config still takes precedence

AlexanderAshitkin reviewed (approved):
I approve this change because it does not introduce any regressions and indeed, could fix certain scenarios. However, I would prefer the following enhancements:

1. An integration test that verifies the fix. This should include scenarios such as: 
   - No configuration set
   - No plugin configuration set
   - If plugin reconciliation set, the default is overridden (or merged)

2. Although this is definitely a step in the right direction, I would love to see some ground work to make this configuration extensible.

To achieve this, I suggest the following:

1. For each plugin, export its all parameters to xml (per plugin).
2. For the exported parameters, annotate each one as either "behavioral" or "functional." (further categorization might also help). For example, the number of compiler threads or log level would be considered behavioral parameters, whereas the target level would be functional.

The goal is to create a closed and matching set of parameters for default configurations for every plugin, every goal. The current implementation does not enforce the categorization of all parameters, which can lead to incorrect behavior if the plugin is modified (for instance, if a critical parameter is renamed).

Ideally, the default configuration should fail on encountering an unknown parameter or at least emit a warning.

cowwoc committed:
Add integration tests and fix default reconciliation merging

- Created integration tests for three scenarios:
  * No executionControl configuration (defaults apply automatically)
  * executionControl with other plugin (defaults still apply to unconfigured plugins)
  * Explicit plugin config (overrides defaults for that plugin only)

- Implemented parameter export/categorization system:
  * Created plugin-parameters.xsd schema
  * Added maven-compiler-plugin.xml (57 parameters)
  * Added maven-install-plugin.xml (22 parameters)
  * Parameter validation with ERROR/WARN logging
  * Version-specific parameter support with best-match selection

- Implemented XML-based default reconciliation:
  * Created default-reconciliation.xsd schema
  * Created defaults.xml with maven-compiler-plugin and maven-install-plugin
  * Implemented DefaultReconciliationLoader to replace hardcoded defaults

- Updated documentation:
  * Clarified that defaults are built-in (not user-customizable)
  * Added example showing how to override defaults via .mvn/maven-build-cache-config.xml
  * Documented parameter validation and version-specific definitions

AlexanderAshitkin reviewed (changes requested):
This is a good improvement. Please recategorize skip parameters to `functional`, and I’d like to get feedback on the possibility of direct reconciliation using the new plugin descriptors.

cowwoc committed:
Reclassify skip parameters from behavioral to functional

The 'skip' and 'skipMain' parameters directly affect build output:
- skip=true: No bytecode produced (compilation skipped entirely)
- skip=false: Bytecode produced normally

This is a functional difference (affects artifacts), not behavioral
(affects only how the build runs). Updated all three occurrences:
- compile goal: skip parameter
- compile goal: skipMain parameter
- testCompile goal: skip parameter

Addresses reviewer feedback on PR #389.

cowwoc committed:
Implement auto-tracking of functional parameters from plugin XML definitions

Replace static defaults.xml with dynamic auto-generation that tracks ALL
functional parameters from plugin-parameters/*.xml files.

Changes:
- Remove defaults.xml (no longer needed - all plugins have XML definitions)
- Remove DefaultReconciliationLoader class (replaced by auto-generation)
- Add generateReconciliationFromParameters() method to CacheConfigImpl
  that automatically creates reconciliation configs by reading functional
  parameters from plugin parameter definitions
- Auto-generation runs as final fallback after explicit config check

Benefits:
- Single source of truth (plugin-parameters/*.xml defines everything)
- More comprehensive tracking (all functional params, not just 3)
- Easier maintenance (no need to keep defaults.xml in sync)
- Future-proof (new plugins with XML definitions get auto-tracking)

For maven-compiler-plugin:
- Before: Tracked 3 params (source, target, release)
- After: Tracks ALL 15 functional params (source, target, release,
  encoding, debug, debuglevel, optimize, compilerArgs, etc.)

All existing tests pass, including DefaultReconciliationTest which
validates the auto-generation works correctly.

Addresses reviewer feedback on PR #389.

cowwoc committed:
Add test to verify all functional parameters are auto-tracked

Add comprehensive unit test to validate that the auto-generation from
plugin XML definitions tracks ALL functional parameters, not just the
3 that were in the old defaults.xml.

Tests verify:
- maven-compiler-plugin tracks MORE than the original 3 parameters
  (source, target, release) and includes encoding, debug, compilerArgs,
  annotationProcessorPaths, etc.
- maven-install-plugin tracks functional parameters (old defaults.xml
  had ZERO properties listed)
- Behavioral parameters (verbose, fork) are NOT auto-tracked

This test ensures the auto-generation is working as designed and
provides more comprehensive tracking than the old defaults.xml system.

elharo reviewed (changes requested):
need to run mvn spotless:apply

cowwoc committed:
Apply code formatting and upgrade spotless tooling

- Upgrade spotless-maven-plugin from 2.44.5 to 3.0.0
- Upgrade palantir-java-format from 2.56.0 to 2.81.0 for Java 25 support
- Run mvn spotless:apply to format code

Addresses review feedback on PR #389.

AlexanderAshitkin reviewed (changes requested):
Overall, LGTM. There is a small change request to use thread safe map implementation in `PluginParameterLoader`

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/CacheConfigImpl.java:
```diff
@@ -248,27 +249,167 @@ public boolean isLogAllProperties(MojoExecution mojoExecution) {
+            String propertyName = property.getPropertyName();
+
+            if (!goalDef.hasParameter(propertyName)) {
+                LOGGER.error(
```
Fixed

cowwoc committed:
Address PR review comments

- Change LOGGER.error to LOGGER.warn for unknown parameter validation
  Unknown parameters may legitimately occur when users manually correct
  outdated bundled configs, so warning level is more appropriate.

- Use ConcurrentHashMap instead of HashMap in PluginParameterLoader
  Ensures thread safety in multi-threaded Maven builds where the cache
  may be accessed from different worker threads.

Addresses review comments from:
- https://github.com/apache/maven-build-cache-extension/pull/389#pullrequestreview-3402287632
- https://github.com/apache/maven-build-cache-extension/pull/389#pullrequestreview-3402293373

cowwoc committed:
Fix backward compatibility for auto-generated reconciliation configs

This commit addresses test failures caused by strict validation when
restoring caches with missing tracked properties.

Root Cause:
- Auto-generated reconciliation configs track functional parameters
- When restoring a cache created with different tracked properties,
  validation rejected the cache if ANY properties were missing
- This prevented reconciliation checks from running and detecting
  actual parameter changes

Changes:

1. CacheControllerImpl.java (isCachedSegmentPropertiesPresent):
   - Allow cache restore even when some tracked properties are missing
   - Changed from failing fast to logging and proceeding
   - Reconciliation check handles missing properties with defaults
   - Provides backward compatibility when adding new tracked parameters

2. BuildCacheMojosExecutionStrategy.java (isParamsMatched):
   - Add exception handling for missing mojo properties
   - Catch all exceptions, not just IllegalAccessException
   - Treat missing properties as "null" instead of crashing
   - Remove 'final' modifier from currentValue for exception handling

3. CacheConfigImpl.java (generateReconciliationFromParameters):
   - Add debug logging for auto-generation process
   - Changed logging from INFO to DEBUG level
   - Helps troubleshoot reconciliation behavior

Test Results:
- All 83 unit tests pass
- All 25 integration tests pass
- Previously failing tests now pass:
  * DefaultReconciliationTest (2 tests)
  * DefaultReconciliationWithOtherPluginTest
  * MandatoryCleanTest (2 tests)
  * IncrementalRestoreTest
  * ForkedExecutionCoreExtensionTest
  * Issue67Test
  * Issue99Test

Addresses review feedback:
- Implements ERROR→WARN logging change per AlexanderAshitkin comment
- Uses ConcurrentHashMap for thread safety per review comment

cowwoc committed:
Revert Spotless Maven Plugin upgrade

Remove Spotless 3.0.0 upgrade and palantirJavaFormat configuration
to keep the PR focused on the auto-tracking functionality.

philippe-granet commented on src/site/markdown/how-to.md:
```diff
@@ -160,6 +160,172 @@ Add `executionControl/runAlways` section:
+### Default Reconciliation Behavior
+
+The build cache extension automatically tracks certain critical plugin properties by default, even without explicit
+`executionControl` configuration. These defaults are loaded from `default-reconciliation/defaults.xml`:
```
**question:** This file `default-reconciliation/defaults.xml` exist ?

cowwoc replied:
Good catch. The old reference:

```text
default-reconciliation/defaults.xml
```

was stale; the defaults are now derived from the plugin parameter descriptors under:

```text
plugin-parameters/maven-compiler-plugin.xml
plugin-parameters/maven-install-plugin.xml
```

I updated this section to point to those descriptors.

cowwoc committed:
docs: correct default reconciliation descriptor references

cowwoc commented:
@olamy This branch has been rebased onto the latest `master` and force-pushed. I also addressed and resolved the remaining review thread about the stale `default-reconciliation/defaults.xml` documentation reference. GitHub reports the PR as mergeable. Could you please merge it?

cowwoc commented:
@AlexanderAshitkin @elharo I’ve rebased the branch, addressed the remaining documentation comment, and all review threads are now resolved. The outstanding requested-changes reviews appear to be stale at this point. Could you please re-review and approve, or let us know what still needs changes?

@olamy If you agree those older requested-changes reviews are stale, could you please dismiss them or merge if appropriate?

cowwoc commented:
@olamy @slachiewicz Can you please merge this PR? I'm not aware of anything blocking it.

cowwoc committed:
Address PR 389 review comments

cowwoc commented on src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:
```diff
@@ -974,14 +974,18 @@ private boolean isCachedSegmentPropertiesPresent(
                         mojoExecution.getExecutionId());
-                return false;
+                // Don't reject the cache - let reconciliation check handle it
             }
```
Addressed in 25e09ea: cache restoration now rejects cached executions missing tracked properties, forcing a normal rebuild.

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -422,8 +430,18 @@ boolean isParamsMatched(
+                // Catch all exceptions including NullPointerException when property doesn't exist in mojo
+                LOGGER.warn(
+                        "Property '{}' not found in mojo {} - treating as null",
+                        propertyName,
+                        mojo.getClass().getSimpleName());
+                currentValue = "null";
             }
```
Addressed in 25e09ea: unexpected extraction failures are no longer treated as:

```text
null
```

They log and return false, so the cache consistency check falls back to a full build. The existing handling for:

```java
IllegalAccessException
```

remains unchanged.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+        } catch (Exception e) {
+            LOGGER.warn("Failed to load parameter definition for {}: {}", artifactId, e.getMessage(), e);
+            return null;
+        }
```
Addressed in 25e09ea: the descriptor parser now uses Java try-with-resources, so the parameter-definition:

```java
InputStream
```

is closed on every path.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+                        bestMatch.getGoals().size(),
+                        bestMatch.getGoals().values().stream()
+                                .mapToInt(g -> g.getParameters().size())
+                                .sum());
```
Addressed in 25e09ea: the parameter-definition load message now uses this log level:

```text
DEBUG
```

cowwoc commented on src/main/resources/plugin-parameters/maven-compiler-plugin.xml:
```diff
@@ -0,0 +1,327 @@
+          <description>Enable preview features</description>
+        </parameter>
+
+        <!-- BEHAVIORAL: Affect how compilation runs but not the output -->
```
Addressed in 25e09ea: the XML comment now states that:

```text
behavioral
```

parameters are the default and output-affecting exceptions are marked:

```text
FUNCTIONAL
```

cowwoc commented:
@slachiewicz I've addressed the comments. Please re-run the check.

cowwoc commented:
@cstamas Can you please trigger another check?

cowwoc committed:
Address PR 389 review comments

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -401,6 +401,14 @@ boolean isParamsMatched(
             CompletedExecution completedExecution) {
         List<TrackedProperty> tracked = cacheConfig.getTrackedProperties(mojoExecution);
 
```
Addressed in ea9634b by removing the redundant null check. The generated `GoalReconciliation#getReconciles()` initializes the list when null; the cache configuration accessor is already non-null.

cowwoc commented on src/main/resources/plugin-parameters/plugin-parameters.xsd:
```diff
@@ -0,0 +1,68 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           targetNamespace="http://maven.apache.org/PLUGIN-PARAMETERS/1.0.0"
+           xmlns="http://maven.apache.org/PLUGIN-PARAMETERS/1.0.0"
+           elementFormDefault="qualified">
+
+  <xs:element name="plugin" type="PluginType"/>
+
```
Addressed in ea9634b: the schema:

```text
plugin-parameters.xsd
```

now declares a namespaced:

```xml
<plugins>
```

root and supports repeated plugin entries. The versioned test descriptor is namespaced, and schema validation passes.

cowwoc commented on src/site/markdown/how-to.md:
```diff
@@ -169,6 +169,175 @@ Add `executionControl/runAlways` section:
+<?xml version="1.0" encoding="UTF-8"?>
+<plugins>
+  <!-- Parameters for versions 1.0.0 through 2.x -->
+  <plugin xmlns="http://maven.apache.org/PLUGIN-PARAMETERS/1.0.0">
```
Addressed in ea9634b: the version-specific example now declares the plugin-parameters namespace on its:

```xml
<plugins>
```

root.

cowwoc commented on src/test/projects/default-reconciliation-with-other-plugin/.mvn/maven-build-cache-config.xml:
```diff
@@ -0,0 +1,32 @@
+    <executionControl>
+        <reconcile>
+            <plugins>
+                <plugin artifactId="maven-surefire-plugin">
```
Addressed in ea9634b: the fixture now specifies:

```xml
goal="test"
```

for the Maven Surefire plugin.

cowwoc commented on src/site/markdown/how-to.md:
```diff
@@ -169,6 +169,175 @@ Add `executionControl/runAlways` section:
+
+The extension automatically validates reconciliation configurations and logs warnings/errors for:
+
+* **Unknown parameters**: Parameters not defined in the plugin's parameter definition (ERROR level)
```
Addressed in ea9634b: the documentation now uses this log level:

```text
WARN
```

It matches the logging call in `CacheConfigImpl.validateReconciliationConfig()`.

cowwoc commented:
@cstamas Can you please trigger another check?

cowwoc committed:
Fix default plugin parameter reconciliation

cowwoc commented:
@cstamas I fixed the build errors. Please trigger another build.

cstamas commented:
Please rebase against latest master.

cowwoc committed:
Address plugin descriptor review comments

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+            // Single plugin definition
+            definitions.add(parsePluginDefinition(root));
+        } else {
+            // Look for multiple <plugin> elements
+            NodeList pluginNodes = root.getElementsByTagName("plugin");
+            for (int i = 0; i < pluginNodes.getLength(); i++) {
+                Element pluginElement = (Element) pluginNodes.item(i);
```
Addressed in 057770e: plugin discovery now uses `getElementsByTagNameNS("*", "plugin")`, and the new prefixed-namespace regression test verifies the loader.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+        NodeList goalsNodes = pluginElement.getElementsByTagName("goals");
+        if (goalsNodes.getLength() > 0) {
+            Element goalsElement = (Element) goalsNodes.item(0);
+            NodeList goalNodes = goalsElement.getElementsByTagName("goal");
```
Addressed in 057770e: goal elements now use namespace-aware lookups such as `getElementsByTagNameNS("*", "goal")`, so prefixed descriptors parse correctly.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+        NodeList parametersNodes = goalElement.getElementsByTagName("parameters");
+        if (parametersNodes.getLength() > 0) {
+            Element parametersElement = (Element) parametersNodes.item(0);
+            NodeList parameterNodes = parametersElement.getElementsByTagName("parameter");
```
Addressed in 057770e: parameter elements now use namespace-aware lookups such as `getElementsByTagNameNS("*", "parameter")`, including prefixed elements.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+        if (nodes.getLength() > 0) {
+            return nodes.item(0).getTextContent().trim();
+        }
+        return null;
```
Addressed in 057770e: text extraction now uses `getElementsByTagNameNS("*", tagName)`, making it prefix-agnostic.

cowwoc commented on src/main/java/org/apache/maven/buildcache/xml/PluginParameterLoader.java:
```diff
@@ -0,0 +1,271 @@
+            return 0;
+        }
+    }
+
+    /**
+     * Parse plugin parameter definitions from XML.
+     * Supports multiple <plugin> elements in a single file for version-specific definitions.
+     */
+    private java.util.List<PluginParameterDefinition> parseDefinitions(InputStream is, String artifactId)
+            throws Exception {
+        java.util.List<PluginParameterDefinition> definitions = new java.util.ArrayList<>();
+
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        factory.setNamespaceAware(true);
+        DocumentBuilder builder = factory.newDocumentBuilder();
+        Document doc = builder.parse(is);
```
Addressed in 057770e: enabled secure processing, rejected `DOCTYPE` declarations, and disabled external entities, external DTDs, external schema access, and XInclude.

cowwoc commented on src/site/markdown/how-to.md:
```diff
@@ -169,6 +169,175 @@ Add `executionControl/runAlways` section:
+
+**Overriding Defaults:** When you explicitly configure `executionControl` for a plugin, your explicit configuration completely
+overrides the defaults for that plugin. For example, to track only the `release` property for maven-compiler-plugin instead
+of the functional parameters from `plugin-parameters/maven-compiler-plugin.xml`:
```
Addressed in 057770e: clarified that explicit `executionControl` overrides the matching plugin goal only; other goals retain their defaults.

cowwoc commented on src/site/markdown/how-to.md:
```diff
@@ -169,6 +169,175 @@ Add `executionControl/runAlways` section:
+All plugin parameters are categorized into two types:
+
+* **Functional Parameters**: Affect the compiled output or build artifacts (e.g., `source`, `target`, `release`, `encoding`)
+* **Behavioral Parameters**: Affect how the build runs but not the output (e.g., `verbose`, `fork`, `maxmem`, `skip`)
```
Addressed in 057770e: removed `skip` from the generic behavioral-parameter example because it is `functional` for output-affecting plugins such as `maven-install-plugin`.

cowwoc committed:
Clarify functional parameter classification

cowwoc committed:
Correct compiler parameter classifications

cowwoc committed:
Replace parameter types with cache-key metadata

cowwoc commented:
I rebased against the latest master and addressed the latest PR comments.

I also want to point out that I updated the parameter metadata to replace the functional/behavioral annotation with `<cache-key>`.

The name was changed because the old terminology described how a property affects plugin execution, while the actual decision we need is whether changing the property must affect cache reuse. `<cache-key>` states that decision directly and avoids ambiguity around properties such as compiler forking, memory limits, and warning handling.
 
Semantics:

- The default value is `false`.
- `<cache-key>true</cache-key>` means the property participates in the generated cache key/reconciliation. Changing its value causes a cache miss.
- `<cache-key>false</cache-key>` or an omitted annotation means that changing the property does not invalidate the cache.
- A cache miss does not delete existing cache entries.

The terminology migration itself preserved the final classifications: `functional` became `true`, and `behavioral` became `false`.

This latest commit changes `<no-cache>` from `false` to `true` for the following properties:

- `showWarnings` and `showDeprecation` for `compile` and `testCompile`: they can affect the build outcome when combined with `failOnWarning`.
- `fork` for `compile` and `testCompile`: it can select a different compiler process or JDK, affecting output or success.
- `maxmem` and `meminitial` for `compile` and `testCompile`: compiler memory settings can affect compilation success.
- `failOnError` and `failOnWarning` for `compile` and `testCompile`: they directly change the success/failure outcome.
- `forceJavacCompilerUse` for `compile` and `testCompile`: selecting the compiler implementation can affect generated output.
- `staleMillis` for `compile` and `testCompile`: it can change which files are recompiled and therefore the generated files.
- `useIncrementalCompilation` for `compile` and `testCompile`: it can change generated files and the build outcome.
- `installAtEnd` for `install`: on reactor failure, it changes whether artifacts are installed and therefore changes project state.
- `updateReleaseInfo` was removed because it does not seem to be valid parameter of the supported `install` goal.

Please let me know if there are any concerns with the above changes.

cowwoc committed:
Align cache-key test naming and comments

cowwoc commented:
@cstamas I rebased against master. Please retry the build.
